Allow preload to accept a const string.

hpvb opened the pull request:
Allow preload to accept a const string.

In preload() parsing this code will lookup the identifier in the local
constant database. If the identifier corresponds to a string constant
it is used as the path for preload().

Currently this does not work for global constants, only constants
declared in the same class as the preload is happening. We can implement
a full fix too. Maybe we can use this PR to discuss the possibilities.

This (partially) fixes #6798

hpvb commented:
This may actually need a little more work as constants not defined in the current class currently can't be resolved.

reduz commented:
At much it should work for current class only

On Feb 14, 2017 1:34 PM, "Hein-Pieter van Braam" <notifications@github.com>
wrote:

> This may actually need a little more work as constants not defined in the
> current class currently can't be resolved.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/7809#issuecomment-279759969>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z25kVACjhCO1-cZe-FtyJdF3iXoS5ks5rcdeogaJpZM4MAsTh>
> .
>

bojidar-bg commented on modules/gdscript/gd_parser.cpp:
```diff
@@ -386,16 +386,35 @@ GDParser::Node* GDParser::_parse_expression(Node *p_parent,bool p_static,bool p_
+				}
 			}
 			if (!valid) {
 				_set_error("expected string constant as 'preload' argument.");
```
You forgot to update the message here

hpvb replied:
Do you have a suggestion? I figured 'string constant' was still pretty accurate.

bojidar-bg commented on modules/gdscript/gd_parser.cpp:
```diff
@@ -1123,13 +1142,13 @@ GDParser::Node* GDParser::_parse_expression(Node *p_parent,bool p_static,bool p_
 				case OperatorNode::OP_NOT: priority=11; unary=true; break;
 				case OperatorNode::OP_AND: priority=12; break;
 				case OperatorNode::OP_OR: priority=13; break;
-				
```
All of those changes shouldn't have happened, since they are unrelated, but let's say it might pass....

hpvb replied:
I completely didn't notice that. I'll clean up the PR, sorry.

hpvb commented:
@reduz  wouldn't it make sense that it should also work for constants that came in through autoloaded classes?

hpvb committed:
Allow preload to accept a const string.

In preload() parsing this code will lookup the identifier in the local
constant database. If the identifier corresponds to a string constant
it is used as the path for preload().

Currently this does not work for global constants, only constants
declared in the same class as the preload is happening. We can implement
a full fix too. Maybe we can use this PR to discuss the possibilities.

This (partially) fixes #6798
